pyccl/concentration.py

```diff
@@ -0,0 +1,317 @@
+            mdef (:obj:`MassDef`): a mass definition object.
+
+        Returns:
+            bool: True if the mass definition is not compatible with
+                this parametrization. False otherwise.
+        """
+        return False
+
+    def _get_consistent_mass(self, cosmo, M, a, mdef_other):
+        """ Transform a halo mass with a given mass definition into
+        the corresponding mass definition that was used to initialize
+        this object.
+
+        Args:
+            cosmo (:obj:`Cosmology`): A Cosmology object.
+            M (float or array_like): halo mass in units of M_sun.
+            a (float): scale factor.
+            mdef_other (:obj:`MassDef`): a mass definition object.
+
+        Returns:
+            float or array_like: mass according to this object's
+            mass definition.
+        """
+        if mdef_other is not None:
+            M_use = mdef_other.translate_mass(cosmo, M, a, self.mdef)
+        else:
+            M_use = M
+        return M_use
+
+    def get_concentration(self, cosmo, M, a, mdef_other=None):
+        """ Returns the concentration for input parameters.
+
+        Args:
+            cosmo (:obj:`Cosmology`): A Cosmology object.
+            M (float or array_like): halo mass in units of M_sun.
+            a (float): scale factor.
+            mdef_other (:obj:`MassDef`): the mass definition object
+                that defines M.
+
+        Returns:
+            float or array_like: concentration.
+        """
+        M_use = self._get_consistent_mass(cosmo,
+                                          np.atleast_1d(M),
+                                          a, mdef_other)
+
+        c = self.concentration(cosmo, M_use, a)
+        if np.isscalar(M):
+            c = c[0]
+        return c
+
+
+class ConcentrationDiemer15(Concentration):
+    name = 'Diemer15'
+
+    def __init__(self, mdef=None):
+        super(ConcentrationDiemer15, self).__init__("Diemer15",
+                                                    mdef)
+
+    def _default_mdef(self):
+        self.mdef = MassDef(200, 'critical')
+
+    def _setup(self):
+        self.kappa = 1.0
+        self.phi_0 = 6.58
+        self.phi_1 = 1.27
+        self.eta_0 = 7.28
+        self.eta_1 = 1.56
+        self.alpha = 1.08
+        self.beta = 1.77
+
+    def _check_mdef(self, mdef):
+        if (np.fabs(mdef.Delta - 200.) > 1E-4) and \
+           (mdef.rho_type != 'critical'):
+            return True
+        return False
+
+    def concentration(self, cosmo, M, a):
+        M_use = np.atleast_1d(M)
+
+        # Compute power spectrum slope
+        R = mass2radius_lagrangian(cosmo, M_use)
+        lk_R = np.log10(2.0 * np.pi / R * self.kappa)
+        lkmin = np.amin(lk_R-0.05)
+        lkmax = np.amax(lk_R+0.05)
+        logk = np.arange(lkmin, lkmax, 0.01)
+        lpk = np.log10(linear_matter_power(cosmo, 10**logk, a))
+        interp = InterpolatedUnivariateSpline(logk, lpk)
+        n = interp(lk_R, nu=1)
```

Comment:
This block of operations seems expensive. A simple finite diff would be ok maybe? Or I bet we could pull this out of the C-level splines we have. Usually I would not care, but a routine like this will sit deep inside loops over halo masses etc and so is more performance sensitive.

Reply:
Mmmh, so what's happening here is essentially finite differences (except through a spline). Also, for a relatively large array of input masses, this should be computationally more efficient than finite differences around each of those.
I buy the point about making C return the spline derivatives (since this is probably useful for other purposes too), but we don't have infrastructure for that yet. I've left a #TODO and will open an issue.

Reply:
Can we actually test that doing a bunch of masses at once is faster this way?

Reply:
Is this an evaluation of the derivative of sigma(R) with respect to mass?

Reply:
Actually, I misspoke. This will only be faster if the array of masses you pass are sampled faster than delta(log10(mass))=0.03. Happy to change it to a finite differences (I just checked the tests pass either way).

Reply:
Sounds good. Let's go with the simpler python code for now. We can leave the todo to expose the C level spline derivs for another time.

Reply:
simpler python code == finite differences